### PR TITLE
#629 Update and consolidate Windows install instructions

### PR DESCRIPTION
**Pull request by mousetraps**
* Point to new VC++ Build Tools Technical Preview.
* Simplify and consolidate install instructions for all Windows versions.

As described in https://github.com/nodejs/node-gyp/issues/629#issuecomment-153196245, the new VC++ build tools provide a minimal set of C++ dependencies on Windows, so there is no longer a need to install full-blown VS. The Build Tools are currently in Technical Preview, but so far folks on the thread have pretty good success with them so far. Currently it is necessary to install both the Windows 8.1 & Windows 10 SDKs, but at RTM, the more minimal [default install should be sufficient](https://github.com/nodejs/node-gyp/issues/629#issuecomment-172416429).

I also went ahead and generally simplified / consolidated the instructions because some of them were too complex, some of them were too sparse, and the repetition was just plain unnecessary & confusing. Additionally I added a link to our [Node.js Guidelines for Windows](https://github.com/Microsoft/nodejs-guidelines/blob/master/windows-environment.md#compiling-native-addon-modules), which provides additional helpful tips. 

I've tested both the Technical Preview and VS2015 Community instructions on Windows 7, 8, 8.1, and 10 with the following packages. 
* bson
* bufferutil
* kerberos
* node-sass
* node_sqlite3
* phantomjs
* utf-8-validate

cc: @orangemocha, @jasonwilliams200OK, @TheAlphaNerd, @am11, @mluparu

### CODE REVIEW

**Comment by mcollina** [comment]
:+1: for me.

**Comment by am11** [comment]
:+1: for simplification. LGTM.

File: README.md [hunk 1]
```diff
@@ -42,30 +42,17 @@ You will also need to install:
...
+      * Option 1: Install [Visual C++ Build Tools (Technical Preview)](http://go.microsoft.com/fwlink/?LinkId=691132), choose **Custom Install**, and select *both* the Windows 8.1 and Windows 10 SDKs.
+      
+      * Option 2: Install [Visual Studio 2015](https://www.visualstudio.com/products/visual-studio-community-vs) (or modify an existing installation) and select *Common Tools for Visual C++* during setup. This also works with the free Community and Express for Desktop editions.
+     
+      > :bulb: [Windows Vista / 7 only] requires [.NET Framework 4.5.1](http://www.microsoft.com/en-us/download/details.aspx?id=40773)
+    
+    * Install [Python 2.7](https://www.python.org/downloads/) (`v3.x.x` is not supported), and run `npm config set python python2.7` (or see below for further instructions on specifying the proper Python version and path.)
+    * Launch cmd, `npm config set msvs_version 2015`
+
+    If the above steps didn't work for you, please visit [Microsoft's Node.js Guidelines for Windows](https://github.com/Microsoft/nodejs-guidelines/blob/master/windows-environment.md#compiling-native-addon-modules) for additional tips.
```
Comment (orangemocha) [inline README.md]: Maybe keep this paragraph with the bulleted list to make it clear that it still pertains to Windows?

**Comment by orangemocha** [comment]
One minor nit, feel free to take it or leave it. LGTM

**Comment by domenic** [comment]
Do these actually work? Last I checked V8 failed to compile with 2015 every once in a while, especially on x64. The issue is that 2015 is not yet supported in the Chromium build pipeline so it gets broken every once in a while as changes are committed that fail in 2015. And the VC++ Build Tools are definitely not tested.

That said, if our current release works using both these configurations, and we have both of them in CI, then this is nice. But we should not recommend configurations that are not actively tested in CI on every commit.

**Comment by ralphtheninja** [comment]
Nice!

**Comment by TheAlphaNerd** [comment]
@mousetraps thank you so much for putting this together. I'm on the slopes of Tahoe today, but I'll follow these instructions to provision a Windows box tomorrow.

**Comment by rvagg** [comment]
@domenic we have vs2015 as part of our CI set now so it's regularly tested for Node.js releases: https://ci.nodejs.org/job/node-compile-windows/, if the _build_ part gets broken in vs2015 then we'd notice it, so far I'm not aware of anything causing problems with stable V8 as it lands in `master`.

**Comment by domenic** [comment]
@rvagg That sounds like a good start. Do you have VS2015 doing x64 builds in particular? And do you have builds created with the VC++ Build Tools suggested in this PR? Mainly I just don't think we should recommend anything that isn't tested in CI.

**Comment by rvagg** [comment]
@domenic good questions, that I will have to defer to @orangemocha and @joaocgreis on.

/cc @nodejs/build

**Comment by am11** [comment]
@domenic, FWIW, VCBuildTools is packing the same version of cl.exe, link.exe etc. as VS2015 Update 1. So as such VCBuildTools is a subset of VS2015Update1.

**Comment by domenic** [comment]
@am11 yep, which is what worries me. If it were a superset I'd imagine there are no problems, but as a subset it might not be sufficient to build---it isn't yet for V8, if I recall the bugs. Or maybe that's Chrome...

**Comment by orangemocha** [comment]
@domenic good points. We can and should definitely add the VC++ build tools in CI for node builds. I'll give that a try...

**Comment by Sogl** [comment]
:+1:

**Comment by bnoordhuis** [comment]
Reviewers, is this PR good to land?

**Comment by orangemocha** [comment]
Not just yet. We ran into some issues building Node itself with the VC++ build tools and we are trying to figure out a way to fix it in the tools or update the instructions to avoid trouble for those users who also build Node. We should have an update shortly.

**Comment by TheAlphaNerd** [comment]
@orangemocha any movement here? I think I'll be getting a windows box soon and can help if this is blocked

**Comment by orangemocha** [comment]
hey @TheAlphaNerd , nice to hear that you'll be working on Windows and thanks for the offer to help!

So the Visual C++ Build Tools (aka VCBT) have been in Node CI for quite some time but in the process of adding them we discovered an issue with the environment scripts that required some hacking to set up the environment. The issue was fixed in [Update 2](https://www.visualstudio.com/downloads/update2-prereleases), but there is still an issue that prevents from building Node when the repo is on a different drive than the VCBT. Note that this doesn't affect building modules.

So in summary this change is almost ready to go, but it would need at least an updated link to the most recent update, and I would prefer for it to have a disclaimer to continue using full VS for users who also build Node. BUT, another pre-release update is supposed to come out at the end of March with the fix for the drive issue, so at this point it might be easier to just wait until then.

**Comment by TheAlphaNerd** [comment]
@orangemocha thanks for the update!

@mousetraps thanks for the patience. This is really going to help a lot of people when it becomes the standard documentation!!!

**Comment by dalinwilliams** [comment]
@mousetraps 

Thanks for this update! I'm working on getting this to work on [Nano Server](http://www.tomsitpro.com/articles/windows-server-2016-nano-server,2-897.html), the interface-less version of windows server. Other than the VC Build Tools, can you think of anything else to run in order to get node-gyp to work on Nano?

Thanks everyone!

**Comment by bsclifton** [comment]
Changes look good :+1: The only thing I'd suggest would be a troubleshooting section. I think many people (like me) don't have a clean install and MSBuild ends up not resolving things properly. More details about that in https://github.com/nodejs/node-gyp/issues/679#issuecomment-211117703

**Comment by bsclifton** [comment]
@orangemocha @mousetraps - looks like the official release for the Visual C++ Build Tools 2015 happened :smiley: I think the PR needs to be updated (under option 1) to link to it, then this should be ready, right?
https://blogs.msdn.microsoft.com/vcblog/2016/03/31/announcing-the-official-release-of-the-visual-c-build-tools-2015/

**Comment by mousetraps** [comment]
Hey everyone, just updated the readme to point to the new version of the Visual C++ build tools, and also removed the requirement to use a custom install because we've addressed that in the product. Lmk if there's anything else!

**Comment by TheAlphaNerd** [comment]
amazing @mousetraps!

I'll go through with a fresh VM and see if I can get things working following this guide

**Comment by orangemocha** [comment]
Awesome! The VC++ build tools have been used in Node's CI for quite some time now, and @joaocgreis and I have also tested this latest version.

LGTM

File: README.md (hunk 1 above)
Comment (TheAlphaNerd) [inline README.md]: I just finished tried to follow these steps with the following windows / node / npm versions
```
npm ERR! Windows_NT 10.0.14332
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "heapdump"
npm ERR! node v4.4.4
npm ERR! npm  v2.15.1
```

I received the error

```
MSBUILD : error MSB4132: The tools version "2.0" is unrecognized. Available tools versions are "14.0", "4.0".
```

**Comment by TheAlphaNerd** [comment]
To get things working with option 1 I had to run the command `npm set msvs_version 2015`

Should this be added to the instructions? Is it needed for the other options?

**Comment by mousetraps** [comment]
@TheAlphaNerd already there! (might be easier to view the .md rendering than the diff itself https://github.com/mousetraps/node-gyp/tree/629#readme)

**Comment by TheAlphaNerd** [comment]
womp... I lose at reading comprehension once again... I will blame it on reading markdown instead of the rendered output.

LGTM!

**Comment by orangemocha** [comment]
Landed in https://github.com/nodejs/node-gyp/commit/303932a21a1561d4b5f14df8840ee159ff22d2ab, with a slightly reworded commit message.

**Comment by mscdex** [comment]
I'm curious, does this mean that vs2013 will no longer work or be sufficient or ?
